packages/std/src/math/mod.rs

```diff
@@ -50,10 +50,28 @@ mod tests {
     {
     }
 
+    /// A trait that ensures other traits are implemented for our integer types
+    trait IntImpl<'a>:
+        AllImpl<'a>
+        + Shl<u32>
+        + Shl<&'a u32>
+        + ShlAssign<u32>
+        + ShlAssign<&'a u32>
+        + Shr<u32>
+        + Shr<&'a u32>
+        + ShrAssign<u32>
+        + ShrAssign<&'a u32>
+    {
+    }
+
     impl AllImpl<'_> for Uint64 {}
     impl AllImpl<'_> for Uint128 {}
     impl AllImpl<'_> for Uint256 {}
     impl AllImpl<'_> for Uint512 {}
+    impl IntImpl<'_> for Uint64 {}
+    impl IntImpl<'_> for Uint128 {}
+    impl IntImpl<'_> for Uint256 {}
+    impl IntImpl<'_> for Uint512 {}
```

Comment:
Can we remove `impl AllImpl<'_> for Uint512 {}` and friends now? They should be required implicitly through the supertrait.

Reply:
Only if we copy all the bounds from `AllImpl` to `IntImpl`. The way it is right now, `IntImpl` requires that they have `AllImpl` too.

Reply:
I'm referring to line 67-70 in the new code. There we implement

```
    impl AllImpl<'_> for Uint64 {}
    impl AllImpl<'_> for Uint128 {}
    impl AllImpl<'_> for Uint256 {}
    impl AllImpl<'_> for Uint512 {}
```

which requires the Uint* types to implement the `AllImpl` traits. However, line 71-74 should imply that. 

E.g. `impl IntImpl<'_> for Uint64 {}` failes to compile if Uint64 misses `Add<&'a Self>` since IntImpl requires AllImpl which requires `Add<&'a Self>`.

Or am I missing something here?

Reply:
I know what you mean, but that's not how it works. Let's use `Uint64` as an example: Line 71 implies that it implements `AllImpl` and all the `Sh*` traits. If we remove line 67, then it does not implement `AllImpl` anymore and line 71 will fail to compile. `AllImpl` does not get automatically implemented just because all the bounds are implemented.

Reply:
Right you are, amen